### PR TITLE
Use Flashinfer TRT-LLM as Llama 4 compatible MoE backend

### PR DESCRIPTION
**Pull request by b8zhong**
<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.ai to discuss further. -->

## Motivation

Since the FI TRT-LLM backend was already added for blockscale MoE, add it for per-tensor FP8, specifically for Llama 4.
Only for SM100

<!-- Describe the purpose and goals of this pull request. -->

## Modifications

Modify the apply code of MoE methods.

> Generally I only used ModelOpt checkpoints, so I didn't add the code in the normal FP8 file.
> There is some issues with ModelOpt loading on main, merge this in after https://github.com/sgl-project/sglang/pull/10154

<!-- Detail the changes made in this pull request. -->

## Accuracy Tests

Before:
```
python3 benchmark/gsm8k/bench_sglang.py --num-shots 8 --num-questions 1319 --parallel 200
Downloading from https://raw.githubusercontent.com/openai/grade-school-math/master/grade_school_math/data/test.jsonl to /tmp/test.jsonl
/tmp/test.jsonl: 732kB [00:00, 33.3MB/s]                                                                                                                                                                                                          
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1319/1319 [00:39<00:00, 33.54it/s]
Accuracy: 0.935
Invalid: 0.000
Latency: 39.421 s
Output throughput: 3492.464 token/s
```

After:
I set parallel 50 due to some strange server issue on main w/ overlap scheduler, it's not related to this PR.
```
python3 benchmark/gsm8k/bench_sglang.py --num-shots 8 --num-questions 1319 --parallel 50
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1319/1319 [01:13<00:00, 17.95it/s]
Accuracy: 0.931
Invalid: 0.000
Latency: 73.587 s
Output throughput: 1864.289 token/s
```

<!-- If this pull request affects model outputs (e.g., changes to the kernel or model forward code), provide accuracy test results. -->

## Benchmarking and Profiling

Before:
```
python -m sglang.launch_server \
  --model-path /mnt/external-quantized-models/models/nvidia__Llama-4-Maverick-17B-128E-Instruct-FP8 \
  --tp 8 \
  --attention-backend triton \
  --model-loader-extra-config '{
    "enable_multithread_load": true,
    "num_threads": 8
  }' \
  --trust-remote-code \
  --moe-runner-backend triton \
  --quantization modelopt_fp8
```

```
============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 8         
Successful requests:                     64        
Benchmark duration (s):                  53.09     
Total input tokens:                      65536     
Total input text tokens:                 65536     
Total input vision tokens:               0         
Total generated tokens:                  65536     
Total generated tokens (retokenized):    65673     
Request throughput (req/s):              1.21      
Input token throughput (tok/s):          1234.33   
Output token throughput (tok/s):         1234.33   
Total token throughput (tok/s):          2468.67   
Concurrency:                             8.00      
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   6632.70   
Median E2E Latency (ms):                 6634.04   
---------------Time to First Token----------------
Mean TTFT (ms):                          144.41    
Median TTFT (ms):                        143.07    
P99 TTFT (ms):                           162.07    
---------------Inter-Token Latency----------------
Mean ITL (ms):                           6.34      
Median ITL (ms):                         6.34      
P95 ITL (ms):                            6.74      
P99 ITL (ms):                            7.00      
Max ITL (ms):                            16.29     
==================================================
```

After:
```
python -m sglang.launch_server \
  --model-path /mnt/external-quantized-models/models/nvidia__Llama-4-Maverick-17B-128E-Instruct-FP8 \
  --tp 8 \
  --attention-backend triton \
  --model-loader-extra-config '{
    "enable_multithread_load": true,
    "num_threads": 8
  }' \
  --trust-remote-code \
  --moe-runner-backend flashinfer_trtllm \
  --quantization modelopt_fp8
```
```
============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 8         
Successful requests:                     64        
Benchmark duration (s):                  45.07     
Total input tokens:                      65536     
Total input text tokens:                 65536     
Total input vision tokens:               0         
Total generated tokens:                  65536     
Total generated tokens (retokenized):    65692     
Request throughput (req/s):              1.42      
Input token throughput (tok/s):          1453.96   
Output token throughput (tok/s):         1453.96   
Total token throughput (tok/s):          2907.91   
Concurrency:                             7.99      
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   5630.01   
Median E2E Latency (ms):                 5630.28   
---------------Time to First Token----------------
Mean TTFT (ms):                          146.01    
Median TTFT (ms):                        146.60    
P99 TTFT (ms):                           169.93    
---------------Inter-Token Latency----------------
Mean ITL (ms):                           5.36      
Median ITL (ms):                         5.35      
P95 ITL (ms):                            5.76      
P99 ITL (ms):                            5.99      
Max ITL (ms):                            34.35     
==================================================
```

<!-- If this pull request impacts inference speed, provide benchmarking and profiling results. -->

| Metric | Before (`--moe-runner-backend triton`) | After (`--moe-runner-backend flashinfer_trtllm`) | Δ | % Change |
|:--|--:|--:|--:|--:|
| Request Throughput (req/s) | 1.21 | 1.42 | +0.21 | +17.4% |
| Input Token Throughput (tok/s) | 1234.33 | 1453.96 | +219.63 | +17.8% |
| Total Token Throughput (tok/s) | 2468.67 | 2907.91 | +439.24 | +17.8% |
| Mean E2E Latency (ms) | 6632.70 | 5630.01 | −1002.69 | −15.1% |
| Mean TTFT (ms) | 144.41 | 146.01 | +1.60 | +1.1% |
| Mean ITL (ms) | 6.34 | 5.36 | −0.98 | −15.5% |

### CODE REVIEW

File: python/sglang/srt/layers/quantization/fp8.py [hunk 1]
```diff
@@ -1182,56 +1182,126 @@ def apply_with_router_logits(
...
+
+            if has_w2_in_scale:
+                o2_scale = layer.w2_input_scale.to(torch.float32) * w2_weight_scale
+            else:
+                o2_scale = w2_weight_scale
+
+            use_routing_scales_on_input = not (has_w13_in_scale and has_w2_in_scale)
+
+            return trtllm_fp8_per_tensor_scale_moe(
+                routing_logits=router_logits.to(torch.float32),
```
[the remark this answers is not in the document]
Comment (b8zhong) [reply python/sglang/srt/layers/quantization/fp8.py]: Remove the code in normal FP8 method

File: python/sglang/srt/layers/quantization/fp8.py [hunk 2]
```diff
@@ -1182,56 +1182,126 @@ def apply_with_router_logits(
...
+                o1_scale = layer.w13_input_scale.to(torch.float32) * w13_weight_scale
+                o1_gate_scale = o1_scale
+            else:
+                o1_scale = w13_weight_scale
+                o1_gate_scale = o1_scale
+
+            if has_w2_in_scale:
+                o2_scale = layer.w2_input_scale.to(torch.float32) * w2_weight_scale
+            else:
+                o2_scale = w2_weight_scale
```
[the remark this answers is not in the document]
Comment (b8zhong) [reply python/sglang/srt/layers/quantization/fp8.py]: Remove the code in normal FP8 method

File: python/sglang/srt/layers/quantization/modelopt_quant.py [hunk 3]
```diff
@@ -481,6 +482,91 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
...
+                    w13_shuffled.append(
+                        shuffle_matrix_a(
+                            w13_interleaved[i].view(torch.uint8), epilogue_tile_m
+                        )
+                    )
+                    w2_shuffled.append(
+                        shuffle_matrix_a(
+                            layer.w2_weight[i].view(torch.uint8), epilogue_tile_m
+                        )
+                    )
```
[the remark this answers is not in the document]
Comment (b8zhong) [reply python/sglang/srt/layers/quantization/modelopt_quant.py]: Done

File: python/sglang/srt/layers/quantization/modelopt_quant.py [hunk 4]
```diff
@@ -492,6 +578,102 @@ def apply(
...
+                not topk_config.num_expert_group
+            ), "ModelOpt FP8 MoE does not support expert grouping"
+            assert (
+                not topk_config.topk_group
+            ), "ModelOpt FP8 MoE does not support grouped top-k"
+            routing_method_type = 3  # Llama4 routing
+
+            print_info_once(
+                f"ModelOptFp8MoEMethod.apply: entering FlashInfer TRTLLM path (top_k={int(topk_config.top_k)}, n_group=0, topk_group=0)"
+            )
```
[the remark this answers is not in the document]
Comment (b8zhong) [reply python/sglang/srt/layers/quantization/modelopt_quant.py]: done + remove print

**Comment by trevor-m** [comment]
@kaixih @pavanimajety  Can you help review?

File: python/sglang/srt/layers/quantization/modelopt_quant.py [hunk 5]
```diff
@@ -481,6 +481,77 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
                 layer.w2_input_scale.max(), requires_grad=False
             )
 
+        # Align FP8 weights to FlashInfer per-tensor kernel layout if enabled
+        if should_use_flashinfer_trtllm_moe():
+            from flashinfer import reorder_rows_for_gated_act_gemm, shuffle_matrix_a
+
+            # 1) Swap W13 halves: [Up, Gate] -> [Gate, Up] expected by FI
```
Comment (kaixih) [inline python/sglang/srt/layers/quantization/modelopt_quant.py]: can we make sure this is not a duplicate step from https://github.com/sgl-project/sglang/blob/7e6191c098e9ff610f9f9f20376bad9071908dc9/python/sglang/srt/layers/moe/fused_moe_triton/layer.py#L609?

It seems we already do the swap during the weight loading instead of the process_weights_after_loading.

File: python/sglang/srt/server_args.py [hunk 6]
```diff
@@ -1150,7 +1162,9 @@ def _handle_moe_kernel_config(self):
 
         if self.moe_runner_backend == "flashinfer_trtllm":
             assert (
-                self.quantization == "modelopt_fp4" or self.quantization == "fp8"
+                self.quantization == "modelopt_fp4"
+                or self.quantization == "modelopt_fp8"
+                or self.quantization == "fp8"
             ), "modelopt_fp4 or fp8 quantization is required for Flashinfer TRTLLM MoE"
```
Comment (kaixih) [inline python/sglang/srt/server_args.py]: nit: modelopt_[fp4|fp8] or fp8 quantization is required for Flashinfer TRTLLM MoE

File: python/sglang/srt/layers/quantization/modelopt_quant.py [hunk 7]
```diff
@@ -526,6 +596,95 @@ def apply(
         layer: torch.nn.Module,
         dispatch_output: StandardDispatchOutput,
     ) -> CombineInput:
+        from sglang.srt.layers.moe.token_dispatcher import StandardCombineInput
+        from sglang.srt.layers.moe.topk import TopKOutputChecker
```
Comment (kaixih) [inline python/sglang/srt/layers/quantization/modelopt_quant.py]: nit: can we put these imports to the top?

File: python/sglang/srt/layers/quantization/modelopt_quant.py [hunk 8]
```diff
@@ -526,6 +596,95 @@ def apply(
...
+            from flashinfer.fused_moe import trtllm_fp8_per_tensor_scale_moe
+
+            correction_bias = (
+                None
+                if topk_config.correction_bias is None
+                else topk_config.correction_bias
+            )
+
+            # Pre-quantize activations to FP8 per-tensor using provided input scale
+            from sglang.srt.layers.quantization.fp8_kernel import scaled_fp8_quant
```
Comment (kaixih) [inline python/sglang/srt/layers/quantization/modelopt_quant.py]: same here.

File: python/sglang/srt/layers/quantization/modelopt_quant.py [hunk 9]
```diff
@@ -526,6 +596,95 @@ def apply(
...
+                and layer.output1_scales_scalar is not None
+            )
+            assert (
+                hasattr(layer, "output1_scales_gate_scalar")
+                and layer.output1_scales_gate_scalar is not None
+            )
+            assert (
+                hasattr(layer, "output2_scales_scalar")
+                and layer.output2_scales_scalar is not None
+            )
```
Comment (kaixih) [inline python/sglang/srt/layers/quantization/modelopt_quant.py]: nit: do we need to assert these again? I am thinking these already created in the above loading process under the same (or a more relaxed) condition.

File: python/sglang/srt/server_args.py (hunk 6 above)
Comment (b8zhong) [reply python/sglang/srt/server_args.py]: Ah I think you reviewed before I rebased & didn't publish comment, I did make this change btw

File: python/sglang/srt/layers/quantization/modelopt_quant.py (hunk 7 above)
Comment (b8zhong) [reply python/sglang/srt/layers/quantization/modelopt_quant.py]: Done

File: python/sglang/srt/layers/quantization/modelopt_quant.py (hunk 8 above)
Comment (b8zhong) [reply python/sglang/srt/layers/quantization/modelopt_quant.py]: Done

File: python/sglang/srt/layers/quantization/modelopt_quant.py (hunk 9 above)
Comment (b8zhong) [reply python/sglang/srt/layers/quantization/modelopt_quant.py]: Hmmm yeah they should be created anyway, I think the assert is just bloating the code. Removed

File: python/sglang/srt/layers/quantization/modelopt_quant.py (hunk 8 above)
Comment (b8zhong) [reply python/sglang/srt/layers/quantization/modelopt_quant.py]: Moved it back bc of some circular import issue

Comment (b8zhong) [reply python/sglang/srt/layers/quantization/modelopt_quant.py]: (from sglang.srt.layers.quantization.fp8_kernel import scaled_fp8_quant is fine acc), I mean the moe imports I moved back to locally)

File: python/sglang/srt/layers/quantization/modelopt_quant.py [hunk 10]
```diff
@@ -526,6 +603,79 @@ def apply(
...
+
+            # Enforce Llama4 routing for ModelOpt FP8 MoE
+            assert topk_config.top_k == 1, "ModelOpt FP8 MoE requires top_k==1"
+            assert (
+                not topk_config.num_expert_group
+            ), "ModelOpt FP8 MoE does not support expert grouping"
+            assert (
+                not topk_config.topk_group
+            ), "ModelOpt FP8 MoE does not support grouped top-k"
+            routing_method_type = 3  # Llama4 routing
```
Comment (kaixih) [inline python/sglang/srt/layers/quantization/modelopt_quant.py]: Can we do sth like: https://github.com/sgl-project/sglang/blob/f80371ff8cb7d7308aea818e54d68de464972856/python/sglang/srt/layers/moe/fused_moe_triton/layer.py#L1073

Comment (b8zhong) [reply python/sglang/srt/layers/quantization/modelopt_quant.py]: Ya, done

**Review by kaixih — APPROVED** [review]

File: python/sglang/srt/layers/quantization/modelopt_quant.py [hunk 11]
```diff
@@ -526,6 +603,81 @@ def apply(
...
+        x = dispatch_output.hidden_states
+        topk_output = dispatch_output.topk_output
+
+        # Fast path: TRT-LLM FP8 per-tensor MoE using BYPASSED TopK routing
+        from sglang.srt.layers.moe.topk import TopKOutputChecker
+
+        if should_use_flashinfer_trtllm_moe() and TopKOutputChecker.format_is_bypassed(
+            topk_output
+        ):
+            router_logits = topk_output.router_logits
```
Comment (Fridge003) [inline python/sglang/srt/layers/quantization/modelopt_quant.py:615]: Can we wrap this part of codes into `self.runner.run` ?

Comment (b8zhong) [reply python/sglang/srt/layers/quantization/modelopt_quant.py:615]: (talked on slack) Due to some circular issue we decide to do this seperately

```
model_config -> layers.quantization.__init__ -> modelopt_quant -> moe.token_dispatcher/topk -> eplb -> model_loader -> model_config
```

**Comment by b8zhong** [comment]
CI failure looks unrelated, do you think it's ready @Fridge003

**Review by Fridge003 — APPROVED** [review]
